plugin/helpers_channels.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package plugin
+
+import (
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/utils"
+	"github.com/pkg/errors"
+)
+
+func (p *HelpersImpl) EnsureChannel(channel *model.Channel) (retChanId string, retErr error) {
+	// Must provide a channel with a name and teadId
+	if channel == nil || len(channel.Name) < 1 || len(channel.TeamId) < 1 {
+		return "", errors.New("passed a bad channel, nil or no name or no team id")
+	}
+
+	// If we fail for any reason, this could be a race between creation of channel and
+	// retrieval from another EnsureChannel. Just try the basic retrieve existing again.
+	defer func() {
+		if retChanId == "" || retErr != nil {
+			var err error
+			var chanIdBytes []byte
+
+			err = utils.ProgressiveRetry(func() error {
+				chanIdBytes, err = p.API.KVGet(CHANNEL_KEY)
+				if err != nil {
+					return err
+				}
+				return nil
+			})
+
+			if err == nil && chanIdBytes != nil {
+				retChanId = string(chanIdBytes)
+				retErr = nil
+			}
+		}
+	}()
+
+	// Fetch channel ID from key value store
+	chanIdBytes, kvGetErr := p.API.KVGet(CHANNEL_KEY)
+	if kvGetErr != nil {
+		// Failed to retrive the value of channel
+		return "", errors.Wrap(kvGetErr, "failed to get channel")
+	}
+
+	// Check if Channel ID exists
+	if chanIdBytes != nil {
+		chanId := string(chanIdBytes)
+		return chanId, nil
+	}
+
+	// Check if channel already exists (ignore deleted channels)
+	if existingChannel, chanGetErr := p.API.GetChannelByName(channel.TeamId, channel.Name, false); chanGetErr == nil && existingChannel != nil {
+		// Update metadata of the channel
+		if updateErr := updateChannelMeta(existingChannel, channel); updateErr != nil {
+			return "", errors.Wrap(updateErr, "Failed to update the metadata of existing channel")
+		}
+
+		// Send the updates to API
+		updatedChannel, chanUpdateErr := p.API.UpdateChannel(existingChannel)
```

Comment:
One minor optimization here can be that I only call `UpdateChannel` function if the metadata was actually updated.